1488 Endo Passable

turadg opened the pull request:
#endo-branch: master 

refs: #1488

## Description

Adapt to actual Passable type

Reviewers, the bar as yet is not whether to merge to this repo. This PR is to serve as evidence that  https://github.com/endojs/endo/pull/2238 can merge and release and agoric-sdk can quickly adopt the new Endo packages.

Once the required Endo code is available by NPM, 
1. bump the Endo packages here to their new releases
2. remove `#endo-branch` pragma
3. verify the Cloudflare job starts passing (it fails now because it doesn't know about #endo-branch)


### Security Considerations

<!-- Does this change introduce new assumptions or dependencies that, if violated, could introduce security vulnerabilities? How does this PR change the boundaries between mutually-suspicious components? What new authorities are introduced by this change, perhaps by new API calls? 
-->

### Scaling Considerations

<!-- Does this change require or encourage significant increase in consumption of CPU cycles, RAM, on-chain storage, message exchanges, or other scarce resources? If so, can that be prevented or mitigated? -->

### Documentation Considerations

<!-- Give our docs folks some hints about what needs to be described to downstream users.

Backwards compatibility: what happens to existing data or deployments when this code is shipped? Do we need to instruct users to do something to upgrade their saved data? If there is no upgrade path possible, how bad will that be for users?

-->

### Testing Considerations

To test locally,
```
cd $ENDO_REPO
git checkout 1488-Passable-redux
git pull
cd $AGORIC_REPO
scripts/replace-package.sh $ENDO_REPO
```


### Upgrade Considerations

<!-- What aspects of this PR are relevant to upgrading live production systems, and how should they be addressed? -->

dckc commented on packages/ERTP/src/amountMath.js:
```diff
@@ -77,9 +80,14 @@ const helpers = {
 /**
- * @template {AmountValue} V
- * @type {(value: V) => AssetKindForValue<V>}
+ * @type {{
+ *   (value: CopyBag): 'copyBag';
+ *   (value: CopySet): 'copySet';
+ *   (value: Array): 'copySet';
+ *   (value: bigint): 'nat';
+ * }}
  */
+// @ts-expect-error cast
 const assertValueGetAssetKind = value => {
```
I wonder if this should take an `unknown`.

dckc commented on packages/ERTP/src/amountMath.js:
```diff
@@ -201,26 +208,39 @@ const AmountMath = {
+   * @template {NatValue | CopySet | CopyBag | SetValue} V
+   * @param {B} brand
+   * @param {V} allegedValue
+   * @returns {B extends Brand<'nat'>
+   *     ? NatAmount
```
doesn't `AmountMath.make(aNatBrand, ['a set value'])` falsify this inference?

dckc commented on packages/ERTP/src/amountMath.js:
```diff
@@ -77,9 +80,14 @@ const helpers = {
+ *   (value: bigint): 'nat';
+ * }}
  */
+// @ts-expect-error cast
```
What are we casting to what?

For each `@ts-expect-error`, I'm trying to understand why we're certain that the declared type is correct.

dckc commented on packages/ERTP/src/amountMath.js:
```diff
@@ -246,29 +267,30 @@ const AmountMath = {
     // @ts-expect-error cast
     const assetKind = assertValueGetAssetKind(value);
-    // @ts-expect-error cast (ignore b/c erroring in CI but not my IDE)
+    // @ts-expect-error different subtype
```
I don't understand this comment.

Perhaps it's more clear if I check out this branch and try removing the `@ts-expect-error`...
Well, no, vs-code says it's not needed.
`yarn lint:types` finds 57 errors in 20 files. I wonder what's going on.

dckc reviewed (commented):
hope to discuss my comments so far

turadg commented:
https://github.com/endojs/endo/pull/2238/ merged to master so I updated the `#endo-branch`. I've also rebased onto master and included type coverage before/after measurements.

Once this is green the next steps are to:
- publish latest Endo
- update this PR to include those Endo versions
- get it back to green
- merge

dckc commented:
odd... commit aefc7cd9f78b44eef7963d9577ff3b8616756543 is empty?

turadg commented:
@dckc I changed the base to https://github.com/Agoric/agoric-sdk/tree/integration-endo-master so this can merge. Then @kriskowal 's  PR bumping Endo will include those commits

dckc commented on packages/inter-protocol/src/proposals/core-proposal.js:
```diff
@@ -27,7 +27,7 @@ export const storeInterContractStartKits = async ({
 }) => {
   /**
-   * @param {Promise<MapStore<string, { instance: Instance }>>} storeP
+   * @param {Promise<MapStore<Instance, { instance: Instance }>>} storeP
```
yikes!

dckc commented on packages/SwingSet/src/devices/vat-admin/device-vat-admin.js:
```diff
@@ -156,6 +156,7 @@ export function buildDevice(tools, endowments) {
 
         // D(devices.vatAdmin).createMeter(remaining, threshold) -> meterID
         if (method === 'createMeter') {
+          /** @type {any} */
```
Does this seem correct? If not, should it have XXX?

array destructuring 2 lines down strongly suggests `Array<something>`.
`unknown[]` would be best - i.e. "tsc, please gripe if I have made any assumptions about the contents of this array."

turadg replied:
I time box types improvements in swingset packages because they're usually a rabbit hole

turadg replied:
This was a vestige of some earlier version of `unserialize`. since `argsCapData` is any `args` is too

dckc commented on packages/SwingSet/src/types-external.js:
```diff
@@ -350,9 +355,7 @@ export {};
  * incarnation.
  *
  *
- * @typedef {object} CreateVatResults
- * @property {object} root
- * @property {VatAdminFacet} adminNode
+ * @typedef {{ adminNode: Guarded<VatAdminFacet>, root: object }} CreateVatResults
```
ouch. that seems hard on devex

turadg replied:
that's because the JSDoc is currently on the interface and not the implementation. `VatAdminFacet` is a concrete type so it would be better to document its implementation and then the result would just be typeof the actual facet.

dckc commented on packages/base-zone/src/types.js:
```diff
@@ -30,9 +30,9 @@ export {};
- * @property {<K extends Key, V extends Passable>(
+ * @property {(label: string, options?: StoreOptions) => MapStore<any, any>} mapStore provide a Map-like store named `label` in the zone
+ * @property {<K>(label: string, options?: StoreOptions) => SetStore<K>} setStore provide a Set-like store named `label` in the zone
+ * @property {<K, V>(
```
is this an edit of an earlier edit in this PR?
That suggests reviewing by commit (which I'm doing) is not a good idea.

It also suggests squashing. is that the plan?

turadg replied:
Plan is not to squash. I could squash some if you advise.

dckc replied:
one idea: smush all the commits together and then split them up by package, in topological order. In each package-commit, describe the changes made to that package.

not that important, but it would be kinda ideal

turadg replied:
That's a good idea. I've already done a lot of commit grooming on this PR but I'll try to remember to try that next time

dckc reviewed (approved):
lots of good stuff!

I'm pretty sure the board change to `PassableCap` is a regression, since it includes `Promise`, which can't be used as a key. I almost requested changes over that. But I'll leave it to your discretion / future work.

turadg commented on packages/zoe/tools/types.js:
```diff
@@ -22,7 +22,7 @@ export {};
  * @property {Amount<'nat'>} amountOut
  * The quoted result of trading `amountIn`
- * @property {import('@agoric/time').TimerService} timer
+ * @property {import('@endo/patterns').RemotableObject & import('@agoric/time').TimerService} timer
```
I think I needed this in order to to add to the methods.

turadg commented on packages/smart-wallet/src/offers.js:
```diff
@@ -10,7 +10,7 @@ import '@agoric/zoe/src/zoeService/types-ambient.js';
  *   invitationSpec: import('./invitations.js').InvitationSpec,
  *   proposal: Proposal,
- *   offerArgs?: unknown
+ *   offerArgs?: any
```
because of the nesting of `unknown`, making it not qualify as `Passable`.

dckc replied:
oh! I see.

(thinking out loud): I wonder if `offerArgs?: Passable` would work...
maybe I'll try it after this lands

turadg replied:
it does! I'm force pushing with that update

turadg commented on packages/vats/src/vat-bank.js:
```diff
@@ -348,8 +348,8 @@ const prepareAssetSubscription = zone => {
 /**
  * @template {AssetKind} [K=AssetKind]
  * @typedef {object} AssetIssuerKit
- * @property {ERef<Mint<K>>} [mint]
- * @property {ERef<Issuer<K>>} issuer
+ * @property {Mint<K>} [mint]
+ * @property {Issuer<K>} issuer
```
Yep

dckc commented on packages/SwingSet/src/types-external.js:
```diff
@@ -2,6 +2,11 @@
 export {};
 
+/**
+ * @import {Guarded} from '@endo/exo')
+ * @import {Passable, RemotableObject} from '@endo/pass-style')
```
stray `)`s:

```suggestion
 * @import {Guarded} from '@endo/exo';
 * @import {Passable, RemotableObject} from '@endo/pass-style';
```

turadg replied:
thanks. I grepped for more and cleaned them up

turadg commented:
Landed in the sync PR
